File: CS/App_Code/CriteriaVisitor.cs
using DevExpress.Data.Filtering;
using System;
using System.Linq;

public class CriteriaVisitor : CriteriaPatcherBase {
    private static CriteriaOperator WrapIntoCustomFunction(CriteriaOperator param) {
        return new FunctionOperator(FunctionOperatorType.Custom,
            new ConstantValue("RemoveDiacritics"), (CriteriaOperator)param);
    }

    public static CriteriaOperator Substitute(CriteriaOperator source) {
        var v = source.ToString();
        return new CriteriaVisitor().AcceptOperator(source);
    }

    protected override CriteriaOperator VisitFunction(FunctionOperator theOperator) {
        if(theOperator.OperatorType == FunctionOperatorType.StartsWith ||
            theOperator.OperatorType == FunctionOperatorType.EndsWith ||
            theOperator.OperatorType == FunctionOperatorType.Contains)
            return new FunctionOperator(theOperator.OperatorType,
                CriteriaVisitor.WrapIntoCustomFunction(theOperator.Operands[0]),
                CriteriaVisitor.WrapIntoCustomFunction(theOperator.Operands[1]));
        return base.VisitFunction(theOperator);
    }
}

File: VB/App_Code/CriteriaVisitor.vb
Imports DevExpress.Data.Filtering
Imports System
Imports System.Linq

Public Class CriteriaVisitor
    Inherits CriteriaPatcherBase

    Private Shared Function WrapIntoCustomFunction(ByVal param As CriteriaOperator) As CriteriaOperator
        Return New FunctionOperator(FunctionOperatorType.Custom, New ConstantValue("RemoveDiacritics"), CType(param, CriteriaOperator))
    End Function

    Public Shared Function Substitute(ByVal source As CriteriaOperator) As CriteriaOperator
        Dim v = source.ToString()
        Return (New CriteriaVisitor()).AcceptOperator(source)
    End Function

    Protected Overrides Function VisitFunction(ByVal theOperator As FunctionOperator) As CriteriaOperator
        If theOperator.OperatorType = FunctionOperatorType.StartsWith OrElse theOperator.OperatorType = FunctionOperatorType.EndsWith OrElse theOperator.OperatorType = FunctionOperatorType.Contains Then
            Return New FunctionOperator(theOperator.OperatorType, CriteriaVisitor.WrapIntoCustomFunction(theOperator.Operands(0)), CriteriaVisitor.WrapIntoCustomFunction(theOperator.Operands(1)))
        End If
        Return MyBase.VisitFunction(theOperator)
    End Function
End Class